File: backend/src/routes/skills.js
import { Router } from 'express';
import { requireAuth } from '../middleware/auth.js';
import { skillManager, SKILLS_DIR } from '../services/skills/manager.js';
import fs from 'fs';
import path from 'path';

export const skillsRouter = Router();

skillsRouter.use(requireAuth);

skillsRouter.get('/skills', async (req, res) => {
  try {
    await skillManager.loadSkills();
    const skills = skillManager.getAllSkills(); // Using getAllSkills to include 'instructions' for editing
    res.json(skills);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

skillsRouter.post('/skills', async (req, res) => {
  try {
    const { name, description, instructions } = req.body;
    if (!name || !name.match(/^[a-z0-9-]+$/)) {
      return res.status(400).json({ error: 'Tên kỹ năng chỉ được chứa chữ thường, số và dấu gạch ngang.' });
    }
    
    const skillDir = path.join(SKILLS_DIR, name);
    const filePath = path.join(skillDir, 'SKILL.md');
    if (fs.existsSync(filePath)) return res.status(400).json({ error: 'Kỹ năng này đã tồn tại.' });
    
    fs.mkdirSync(skillDir, { recursive: true });
    const content = `---\nname: "${name}"\ndescription: "${description || ''}"\n---\n\n${instructions || ''}`;
    fs.writeFileSync(filePath, content, 'utf8');
    
    await skillManager.loadSkills();
    res.json({ success: true, name });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

skillsRouter.put('/skills/:name', async (req, res) => {
  try {
    const { name } = req.params;
    const { description, instructions } = req.body;
    
    const filePath = path.join(SKILLS_DIR, name, 'SKILL.md');
    if (!fs.existsSync(filePath)) return res.status(404).json({ error: 'Kỹ năng không tồn tại.' });
    
    const content = `---\nname: "${name}"\ndescription: "${description || ''}"\n---\n\n${instructions || ''}`;
    fs.writeFileSync(filePath, content, 'utf8');
    
    await skillManager.loadSkills();
    res.json({ success: true, name });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

skillsRouter.delete('/skills/:name', async (req, res) => {
  try {
    const { name } = req.params;
    const skillDir = path.join(SKILLS_DIR, name);
    const filePath = path.join(skillDir, 'SKILL.md');
    
    if (fs.existsSync(filePath)) {
      fs.rmSync(skillDir, { recursive: true, force: true });
      await skillManager.loadSkills();
    }
    res.json({ success: true });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

File: backend/api/routers/skills.py
"""Skills router — list, create, update, delete skills (delegates to file system)."""

import json
import os
from pathlib import Path
from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel
from typing import Optional

from hagent_constants import get_hagent_home, _APP_ROOT

router = APIRouter(prefix="/api/skills", tags=["skills"])


def _skills_dir() -> Path:
    return _APP_ROOT / "skills"


def _build_skill_md(name: str, description: str, instructions: str) -> str:
    return f"""---
name: "{name}"
description: "{description or ''}"
---

{instructions or ''}
"""


def _parse_skill_md(path: Path) -> dict:
    content = path.read_text("utf-8")
    name = path.parent.name
    desc = ""
    instructions = content
    if content.startswith("---"):
        parts = content.split("---", 2)
        if len(parts) >= 3:
            for line in parts[1].strip().split("\n"):
                if line.startswith("description:"):
                    desc = line.split(":", 1)[1].strip().strip('"')
            instructions = parts[2].strip() if len(parts) > 2 else ""
    return {"name": name, "description": desc, "instructions": instructions, "file": str(path)}


@router.get("")
def list_skills(request: Request):
    skills_dir = _skills_dir()
    if not skills_dir.exists():
        return []
    skills = []
    for skill_md in sorted(skills_dir.rglob("SKILL.md")):
        rel = skill_md.relative_to(skills_dir).parent
        name = "/".join(rel.parts)
        parsed = _parse_skill_md(skill_md)
        parsed["name"] = name
        skills.append(parsed)
    return skills


class CreateSkillBody(BaseModel):
    name: str
    description: Optional[str] = ""
    instructions: Optional[str] = ""


@router.post("")
def create_skill(body: CreateSkillBody, request: Request):
    import re
    if not re.match(r"^[a-z0-9-]+$", body.name):
        raise HTTPException(status_code=400, detail="Ten ky nang chi chua chu thuong, so va gach ngang.")
    skill_dir = _skills_dir() / body.name
    skill_md = skill_dir / "SKILL.md"
    if skill_md.exists():
        raise HTTPException(status_code=400, detail="Ky nang nay da ton tai.")
    skill_dir.mkdir(parents=True, exist_ok=True)
    skill_md.write_text(_build_skill_md(body.name, body.description or "", body.instructions or ""), "utf-8")
    return {"success": True, "name": body.name}


class UpdateSkillBody(BaseModel):
    description: Optional[str] = ""
    instructions: Optional[str] = ""


@router.put("/{name}")
def update_skill(name: str, body: UpdateSkillBody, request: Request):
    skill_md = _skills_dir() / name / "SKILL.md"
    if not skill_md.exists():
        raise HTTPException(status_code=404, detail="Ky nang khong ton tai.")
    skill_md.write_text(
        _build_skill_md(name, body.description or "", body.instructions or ""), "utf-8"
    )
    return {"success": True, "name": name}


@router.delete("/{name}")
def delete_skill(name: str, request: Request):
    import shutil
    skill_dir = _skills_dir() / name
    if skill_dir.exists():
        shutil.rmtree(skill_dir)
    return {"success": True}
